src/backends/plonky2/circuits/mainpod.rs

```diff
@@ -137,91 +137,211 @@ impl OperationVerifyGate {
-        let is_new_entry = builder.is_equal(native_op, op_new_entry);
-        op_flags.push(is_new_entry);
-        let op_copy_statement =
-            builder.constant(F::from_canonical_u64(NativeOperation::CopyStatement as u64));
-        let is_copy_statement = builder.is_equal(native_op, op_copy_statement);
-        op_flags.push(is_copy_statement);
-        let op_eq_from_entries = builder.constant(F::from_canonical_u64(
-            NativeOperation::EqualFromEntries as u64,
-        ));
-        let is_eq_from_entries = builder.is_equal(native_op, op_eq_from_entries);
-        op_flags.push(is_eq_from_entries);
-        let op_lt_from_entries =
-            builder.constant(F::from_canonical_u64(NativeOperation::LtFromEntries as u64));
-        let is_lt_from_entries = builder.is_equal(native_op, op_lt_from_entries);
-        op_flags.push(is_lt_from_entries);
-        let op_not_contains_from_entries = builder.constant(F::from_canonical_u64(
-            NativeOperation::NotContainsFromEntries as u64,
-        ));
-        let is_not_contains_from_entries =
-            builder.is_equal(native_op, op_not_contains_from_entries);
-        op_flags.push(is_not_contains_from_entries);
-
-        // One supported operation must be used.  We sum all operation flags and expect the result
-        // to be 1.  Since the flags are boolean and at most one of them is true the sum is
-        // equivalent to the OR.
-        let or_op_flags = op_flags
-            .iter()
-            .map(|b| b.target)
-            .fold(_false.target, |acc, x| builder.add(acc, x));
-        builder.connect(or_op_flags, _true.target);
+        // Op checks to carry out. Each 'eval_X' should be thought of
+        // as 'eval' restricted to the op of type X, where the
+        // returned target is `false` if the input targets lie outside
+        // of the domain.
+        let op_checks = vec![
+            vec![
+                self.eval_none(builder, st, op),
+                self.eval_new_entry(builder, st, op, prev_statements),
+            ],
+            // Skip these if there are no resolved op args
+            if resolved_op_args.len() == 0 {
+                vec![]
+            } else {
+                vec![
+                    self.eval_copy(builder, st, op, &resolved_op_args)?,
+                    self.eval_eq(builder, st, op, &resolved_op_args),
+                    self.eval_lt(builder, st, op, &resolved_op_args),
+                ]
+            },
+        ]
+        .concat();
 
-        let ok = builder._true();
-        let none_ok = self.eval_none(builder, st, op);
-        let ok = builder.select_bool(is_none, none_ok, ok);
-        let new_entry_ok = self.eval_new_entry(builder, st, op);
-        let ok = builder.select_bool(is_new_entry, new_entry_ok, ok);
+        let ok = builder.any(op_checks);
 
         builder.connect(ok.target, _true.target);
 
         Ok(OperationVerifyTarget {})
     }
 
+    fn eval_eq(
```

Comment:
did you consider naming this `eval_eq_from_entries` so that it's clear this is checking an operation and not a statement?